sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -520,6 +523,25 @@ class Dataset[T] private[sql](
     }
   }
 
+  /**
+   * Using inner equi-join to join this [[Dataset]] returning a [[Tuple2]] for each pair
+   * where `condition` evaluates to true
+   *
+   * @since 1.6.0
+   */
+  def joinWith[U](other: Dataset[U], condition: Column): Dataset[(T, U)] = {
+    joinWith(other, condition, "inner")
+  }
+
+  /**
+   * Joins this [[Dataset]] returning a [[Tuple2]] for each pair using cartesian join
+   *
+   * Note that cartesian joins are very expensive without an extra filter that can be pushed down.
+   *
+   * @since 1.6.0
+   */
+  def joinWith[U](other: Dataset[U]): Dataset[(T, U)] = joinWith (other, lit(true), "inner")
```

Comment:
Actually I'd maybe just remove this for now -- since cartesian joins are too expensive.